RecentActivity/src/org/sleuthkit/autopsy/recentactivity/MessageURLAnalyzer.java

```diff
@@ -0,0 +1,488 @@
+                .filter(StringUtils::isNotBlank)
+                .collect(Collectors.toList());
+
+        MessageDomainTrieNode node = rootTrie;
+        // the root node is null indicating we won't be able to do a lookup.
+        if (node == null) {
+            return null;
+        }
+
+        // iterate through tokens in reverse order
+        int idx = tokens.size() - 1;
+        for (; idx >= 0; idx--) {
+            node = node.getChild(tokens.get(idx));
+            // if we hit a leaf node or we have no matching child node, continue.
+            if (node == null || node.getMessageType() != null) {
+                break;
+            }
+        }
+
+        MessageType messageType = node != null ? node.getMessageType() : null;
+
+        if (messageType == null) {
+            return null;
+        } else {
+            // if there is a message type, we have a result.  Concatenate the 
+            // appropriate domain tokens and return.
+            int minIndex = Math.max(0, idx);
+            List<String> subList = tokens.subList(minIndex, tokens.size());
+            String hostSuffix = String.join(JOINER, subList);
+            return Pair.of(hostSuffix, messageType);
+        }
+    }
+
+    /**
+     * Goes through web history artifacts and attempts to determine any hosts of
+     * a message type. If any are found, a TSK_DOMAIN_CATEGORY artifact is
+     * created (at most one per host suffix).
+     */
+    private void findMessageDomains() {
+        if (this.rootTrie == null) {
+            logger.log(Level.SEVERE, "Not analyzing message domain.  No root trie loaded.");
+            return;
+        }
+
+        int artifactsAnalyzed = 0;
+        int messageDomainInstancesFound = 0;
+
+        // only one suffix per ingest is captured so this tracks the suffixes seen.
+        Set<String> domainSuffixesSeen = new HashSet<>();
+
+        try {
+            Collection<BlackboardArtifact> listArtifacts = currentCase.getSleuthkitCase().getBlackboard().getArtifacts(
+                    Arrays.asList(new BlackboardArtifact.Type(ARTIFACT_TYPE.TSK_WEB_HISTORY)),
+                    Arrays.asList(dataSource.getId()));
+
+            logger.log(Level.INFO, "Processing {0} blackboard artifacts.", listArtifacts.size()); //NON-NLS
+
+            for (BlackboardArtifact artifact : listArtifacts) {
+                // make sure we haven't cancelled
+                if (context.dataSourceIngestIsCancelled()) {
+                    break;       //User cancelled the process.
+                }
+
+                // make sure there is attached file
+                AbstractFile file = tskCase.getAbstractFileById(artifact.getObjectID());
+                if (file == null) {
+                    continue;
+                }
+
+                // get the url string from the artifact
+                BlackboardAttribute urlAttr = artifact.getAttribute(new BlackboardAttribute.Type(BlackboardAttribute.ATTRIBUTE_TYPE.TSK_URL));
```

Comment:
Why not get the TSK_DOMAIN attribute that will be on this artifact? Then, you could skip all the business about getting the host.

Reply:
Reviewing some code, this extractor will be run after the browser extractors.